core/src/main/scala/io/projectglow/sql/expressions/LogisticRegressionGwas.scala

```diff
@@ -175,27 +179,33 @@ case class NewtonResult(
     converged: Boolean,
     exploded: Boolean)
 
+/** Base trait for logistic regression tests */
 trait LogitTest extends Serializable {
+
+  /**
+   * Type for the state that the test maintains between variants.
+   */
   type FitState
+
   def resultSchema: StructType
-  def canReuseNullFit: Boolean
-  def fitNullModel(phenotypes: Array[Double], covariates: SparkDenseMatrix): FitState
-  def runTest(
-      x: DenseMatrix[Double],
-      y: DenseVector[Double],
-      nullModelFit: Option[FitState]): InternalRow
-}
 
-trait LogitTestWithoutNullFit extends LogitTest {
-  final override val canReuseNullFit: Boolean = false
+  /**
+   * If true, the [[FitState]] uses per-phenotype information and must be refit for each phenotype.
+   * If false, the [[FitState]] only depends on the `covariates`
+   * @return
+   */
+  def fitStatePerPhenotype: Boolean
 
-  final override def fitNullModel(
-      phenotypes: Array[Double],
-      covariates: SparkDenseMatrix): FitState = {
-    throw new IllegalArgumentException("Test does not support null model fit")
-  }
-}
+  /**
+   * Initializes a [[FitState]] for a (covariate matrix, phenotype array) pair.
+   *
+   * As much memory allocation as possible should be performed in this step to avoid allocations
+   * in the per-row fit.
+   */
+  def init(phenotypes: Array[Double], covariates: SparkDenseMatrix): FitState
```

Comment:
If fit state per phenotype is false, then `phenotypes` will not be used in the initialization, correct?

Reply:
True